### ISSUE TITLE
refleak in _JellyableAvatarMixin

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5079]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)| @dustin reported|
|-|-|
|Trac ID|trac#5079|
|Type|enhancement|
|Created|2011-05-04 04:11:25Z|
|Branch|https://github.com/twisted/twisted/tree/pb-logout-leak-5079|

We have PB clients disconnect unexpectedly quite often.  Networks fail a lot!  We noticed that ```gc.garbage``` gains a RemoteReference for each such disconnection - indicating a slow but persistent leak.

Chasing this down, we have a ```requestAvatar``` method that creates a perspective and then returns ```(pb.IPerspective, persp, lambda: persp.detached(mind))```.  Note, in particular, that the logout callable maintains a reference to the perspective.

Here's where we see a problem:
```python
    def _cbLogin(self, (interface, avatar, logout)):
        """  
        Ensure that the avatar to be returned to the client is jellyable and
        set up disconnection notification to call the realm's logout object.
        """
        if not IJellyable.providedBy(avatar):
            avatar = AsReferenceable(avatar, "perspective")

        puid = avatar.processUniqueID()

        def dereferenceLogout():
            self.broker.dontNotifyOnDisconnect(logout)
            logout()

        self.broker._localCleanup[puid] = dereferenceLogout
        # No special helper function is necessary for notifyOnDisconnect
        # because dereference callbacks won't be invoked if the connection is
        # randomly dropped.  I'm not sure those are ideal semantics, but this
        # is the only user of the (private) API at the moment and it works just
        # fine as things are. -exarkun
        self.broker.notifyOnDisconnect(logout)
        return avatar
```

Here, ```dereferenceLogout``` maintains a reference to ```logout```, which references the perspective, which references ```dereferenceLogout```.  At some point in this cycle, there's a RemoteReference, and that has a ```__del__``` method, making it uncollectable.

A fix that seems to work is to contradict exarkun's comment above and add a special helper function to ```notifyOnDisconnect``` that drops the ```_localCleanup``` entry for ```dereferenceLogout```.  Our uses don't ever invoke ```dereferenceLogout```, but I suspect the correct solution is to remove *both* reference in either case.

Does this solution make sense?  If so, I'll post a patch for it.

Attachments:

* [5079-r1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e12/e12809c9823bd87c393ef4771deedde51bf314b2/d165986a0f8e700ff9fd45c5ee002cd905ae6f30.patch) (6394 bytes) - added by dustin on 2011-05-05 18:41:41Z - 
* [bug5079.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e12/e12809c9823bd87c393ef4771deedde51bf314b2/1cc6aed3ce2a186c44e6dd731e2a4bcbbbee7080.patch) (7507 bytes) - added by dustin on 2011-06-13 02:03:12Z - bug5079-r2.patch
* [bug5079-r3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e12/e12809c9823bd87c393ef4771deedde51bf314b2/e01604228aaca443a4e25d0f526de454d536d32b.patch) (6267 bytes) - added by dustin on 2011-06-29 04:25:17Z - bug5079-r3.patch
* [bug5079-r4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e12/e12809c9823bd87c393ef4771deedde51bf314b2/1b0f5de6d7682095a382a8d0b39250680b485e19.patch) (6352 bytes) - added by dustin on 2011-06-29 04:29:49Z - bug5079-r4.patch
* [bug5079-r5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e12/e12809c9823bd87c393ef4771deedde51bf314b2/53f65c9de9b7c5ce4af5d3675bb9a95cdc94b083.patch) (7071 bytes) - added by dustin on 2011-06-29 05:23:57Z - bug5079-r5.patch

<details><summary>Searchable metadata</summary>

```
trac-id__5079 5079
type__enhancement enhancement
reporter__dustin dustin
priority__normal normal
milestone__ 
branch__branches_pb_logout_leak_5079 branches/pb-logout-leak-5079
branch_author__dustin dustin
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1304482285000000 1304482285000000
changetime__1310129143000000 1310129143000000
version__None None
owner__exarkun exarkun
cc__catlee
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @dustin|
|-|-|

Thanks for tracking this down.  The explanation seems to make sense, I think.  But a unit test demonstrating the problem would be a nice confirmation, too.

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_2"></a>|@dustin commented|
|-|-|

What's the best way to demonstrate a leak in a unit test?  asserting that the length of ```gc.garbage``` doesn't change?

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_3"></a>|@dustin commented|
|-|-|

Here's my first attempt at a patch - you'll not that without the pb.py patch, this fails.

I'm particularly unhappy about the ```os.close()```, but ```loseConnection``` triggers a call to ```_localCleanup[luid]```, not via ```notifyOnDisconnect```, which is where the leak comes from.  If there's a better way to create such an "abrupt" disconnection (in practice, we see it when a slave reboots without warning), I'm all ears.

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_4"></a>|@dustin set owner to @exarkun|
|-|-|

(to exarkun for review, since he commented)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to catlee|
|-|-|

Thanks for reporting this and submitting the patch.  Sorry it took so long to get back to you with the review.

The `os.close` is indeed something to be unhappy about.  It won't work on IOCP reactor, and in general there's no guarantee that the `ITransport` has a `fileno` method that you can treat like that.

Instead, how about a test that doesn't use real connections?  You should be able to construct a PBServerFactory and log in to it directly, using plain old method calls.  Once you have enough state set up, you can call connectionLost on it directly and verify it forgets about the things it is supposed to forget about.

I think it may also be worth considering re-implementing the cleanup behavior using weak references.  Without any `__del__` methods, the cycles are collectable, and the code doesn't have to continue to be insanely cautious about every single reference it creates.  The `RemoteReference` instance itself isn't needed to send the decref, only its luid.  The `Broker` could gain another dictionary (a weak value dict probably) for tracking these and firing the decref messages when needed.

**Comment by twisted-trac** [comment]
|<img alt="catlee's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|catlee set owner to @dustin|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_9"></a>|@dustin commented|
|-|-|

I couldn't actually reproduce this in the test script without the ```os.close``` or real connections - after trying quite a few things.  Were this a regression test for a more reproducible bug, I'd agree, but I don't think I can reproduce this any other way.

How would you feel about conditioning the test on the presence of a ```.fileno``` on the ITransport and not using the IOCP reactor?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

If you don't use a real transport, then you don't need the `os.close` either.  I'm not sure if you're familiar with `twisted.test.proto_helpers.StringTransport`; if not, take a look.  It's a _strength_ of Twisted's protocol/transport separation that you can easily reproduce any network condition on demand, without the cooperation of the network. :)

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_11"></a>|@dustin set owner to @exarkun|
|-|-|

I hooked up a handy pipe for StringTransport, and managed to make a client and server talk over that pipe.  In my testing, without the pb.py patch, the final assertion will fail; with the patch, it passes.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set owner to @dustin|
|-|-|

Thanks for the patch, and thanks for going through another round of review!

Mostly I'm just quoting from the [coding standard](http://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html) here, so please give it a read to save yourself some time in the future.

  1. The `@ivar` declarations on `StringTransportPipe` are wrong; there are
  `leftProtocol` and `rightProtocol` attributes, and they should each be
  documented.
  2. `StringTransportPipe` looks quite a lot like `IOPump`, and also possibly
  like some things in `twisted.protocols.loopback`.  Please re-use one of these
  if you can; the fewer duplicate implementations of this functionality we can
  have, the better.  (It would be _best_ to replace IOPump with something
  based on `proto_helpers`, as `proto_helpers` is emerging as the least ad-hoc
  of all these test helpers, but that might be too much work for this branch.)
  3. There should be 2 blank lines between methods, and 3 between classes, in
  new code.
  4. Nesting classes is generally a bad idea.  `LeakyRealm` should be at module
  scope.
  5. The naming convention for tests (or methods with prefixes in general) is
  `prefix_fooBar`; so, `test_logout_leak` should be `test_logoutLeak`.
  6. No simple suites, please.  In other words, always newline+indent for every
  control structure.  '`if not logout: return`' should have a line break after
  the `:`.
  7. There's no need to add a new import for `defer` in `test_pb`; `Deferred` is
  already imported.
  8. Don't use the 'assert' keyword.  'assert' statements aren't run when Python
  is invoked with `-OO`, and we'd love for Twisted to work (and be tested) in
  that mode as well.
  9. The naming convention for attributes is '`fooBar`', not '`foo_bar`';
  `mind_ref` should be `mindRef`.
  10. Don't assert on the `Deferred.called` attribute.  Whether or not a
  Deferred was called is generally of little interest, unless you're testing
  `Deferred` itself.  You should instead assert about things happening inside
  the callbacks that you have added.  Actually, it's best if you have all of
  your callbacks within tests simply append things to some structure (like a
  list), and then do your asserts on the contents of those structures, inline
  within the test.

Other than these small quibbles it's looking pretty good.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_13"></a>|@dustin commented|
|-|-|

1 and 2 are taken care of by using IOPump.  3 is done (and IOPump updated).  4, 5, 6, 7, 8 (embarassing!), and 9 are easy.  Somehow I missed 10 despite reading the list multiple times.  Short attention span?  Another patch in a moment.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

Thanks again.

  1. The patch should contain a [NEWS file](http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles).
  2. `IOPump.stop` needs a docstring.
  3. `LeakyRealm` could deal with a couple of `@ivar`s, although it would probably be a little easier to explain if it set `mindRef` some other way?  Perhaps passing the mind to a callable and then using `list.append` as said callable?  That would allow the test to ensure that `requestAvatar` was only called once, too.
  4. Would it be possible to have this test method contain less code by using `connectedServerAndClient`?  Maybe I'm missing something but it looks fairly redundant.

That does look like about it, though.

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_16"></a>|@dustin commented|
|-|-|

Replying to [glyph](#note_15):
>   3. `LeakyRealm` could deal with a couple of `@ivar`s, although it would probably be a little easier to explain if it set `mindRef` some other way?  Perhaps passing the mind to a callable and then using `list.append` as said callable?  That would allow the test to ensure that `requestAvatar` was only called once, too.

Sounds good.  I don't think that this test is responsible for counting the number of times `requestAvatar` is called, though.

>   4. Would it be possible to have this test method contain less code by using `connectedServerAndClient`?  Maybe I'm missing something but it looks fairly redundant.

It turns out that the only difference is the realm, and that's easy enough to add as a parametr to `connectedServerAndClient`.  So yes.

> That does look like about it, though. 

I think you should get decreasing points for subsequent reviews of the same ticket ;)

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)|@dustin removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [32258]) Branching to 'pb-logout-leak-5079'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

(In [32259]) Apply bug5079-r5.patch 

refs [#5079](https://github.com/twisted/twisted/issues/5079)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [32260]) and the news fragment

refs [#5079](https://github.com/twisted/twisted/issues/5079)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

(In [32261]) Fix the name of the thing being fixed; change some more wording as long as I'm here anyway.

refs [#5079](https://github.com/twisted/twisted/issues/5079)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/pb-logout-leak-5079)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun set owner to @exarkun|
|-|-|

Okay, basically looks good to me (and buildbot).  One minor thing is the docstring for the test method, which starts with "Test that ..." which we don't like :)  I'll change it to just explain the behavior being tested and then merge.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [32266]) Twiddle the test method docstring

refs [#5079](https://github.com/twisted/twisted/issues/5079)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set status to `closed`|
|-|-|

(In [32267]) Merge pb-logout-leak-5079

Author: dustin
Reviewer: glyph, exarkun
Fixes: [#5079](https://github.com/twisted/twisted/issues/5079)

Ensure that the reference cycle between PB perspective and logout function
is broken when a PB connection is unexpectedly closed so that those objects
do not leak.
